Export bad link and homepage URLs and status

emmabeynon opened the pull request:
[Trello card](https://trello.com/c/SCWYMyWG/572-address-tim-s-llm-data-needs-3)

This adds functionality to export a CSV of non-200 status (aka "bad") link URLs and their status. A similar CSV is available for Local Authority homepages as well. The CSVs are available to download via URLs so that a performance analyst can access the data when required.

emmabeynon committed:
Export bad link and homepage URLs and status

[Trello card](https://trello.com/c/SCWYMyWG/572-address-tim-s-llm-data-needs-3)

This adds functionality to export a CSV of non-200 status (aka "bad") link URLs and their status. A similar CSV is available for Local Authority homepages as well. The CSVs are available to download via URLs so that a performance analyst can access the data when required.

issyl0 commented on spec/lib/local-links-manager/export/bad_links_url_and_status_exporter_spec.rb:
```diff
@@ -0,0 +1,22 @@
+
+  describe ".local_authority_bad_homepage_url_and_status_csv" do
+    it "returns the URL and status for each local authority with a non-200 homepage" do
+      allow(LocalAuthority).to receive_message_chain("where.not.distinct.pluck").and_return([["http://www.hogsmeade.gov.uk", "403"], ["http://www.godricshollow.gov.uk", "Timeout Error"], ["http://www.azkaban.gov.uk/", "Connection failed"], ["http://www.littlehangleton.gov.uk", "404"]])
```
😍 I approve of your test data.

davidbasalla commented on lib/local-links-manager/export/bad_links_url_and_status_exporter.rb:
```diff
@@ -0,0 +1,27 @@
+      def self.bad_links_url_and_status_csv
+        CSV.generate do |csv|
+          csv << HEADINGS
+          Link.enabled_links.where.not(status: "200").distinct.pluck(:url, :status).each do |link|
```
We could use the `currently_broken` Link scope here

davidbasalla replied:
`Link.enabled_links` actually returns a different number than if we looped through each LocalAuthority and called `provided_service_links` on it. The former currently returns '40895' in Integration, the latter '39105'. I think this is due to us not taking tiers into account for `enabled_links`. I've been meaning to have a look at this but not had a chance to. We should look at this at some point in the near future. It's probably fine for this PR, as the difference isn't very large.

emmabeynon replied:
Thanks - this is now using `currently_broken`.

davidbasalla commented on spec/lib/local-links-manager/export/bad_links_url_and_status_exporter_spec.rb:
```diff
@@ -0,0 +1,22 @@
+
+  describe ".local_authority_bad_homepage_url_and_status_csv" do
+    it "returns the URL and status for each local authority with a non-200 homepage" do
+      allow(LocalAuthority).to receive_message_chain("where.not.distinct.pluck").and_return([["http://www.hogsmeade.gov.uk", "403"], ["http://www.godricshollow.gov.uk", "Timeout Error"], ["http://www.azkaban.gov.uk/", "Connection failed"], ["http://www.littlehangleton.gov.uk", "404"]])
```
Stubbing `"where.not.distinct.pluck"` feels a bit brittle. Could we try replacing this with creating some links via FactoryGirl?

emmabeynon replied:
Now using factories and tidied up the expectations to make them more readable 👍

davidbasalla commented on spec/lib/local-links-manager/export/bad_links_url_and_status_exporter_spec.rb:
```diff
@@ -0,0 +1,49 @@
+  let(:exporter) { LocalLinksManager::Export::BadLinksUrlAndStatusExporter }
+
+  before do
+    create(:local_authority, homepage_url: "http://www.hogsmeade.gov.uk", status: "403")
```
💯 for these tests 😄 

A small thing, I'd probably move Link factory calls down to the tests for the links tests (into a separate `before` block). Should we also update 'links_status_exporter_spec.rb' to use FactoryGirl as well while we're at it?

emmabeynon replied:
Thanks!  I've broken the before blocks out and updated links_status_exporter_spec.

emmabeynon committed:
Update Link Status Exporter specs

After creating factories in the Bad Links Url And Status Exporter specs, we have amended the Link Status Exporter specs to adopt the same approach to make the tests less brittle.

davidbasalla commented:
Great! 👍
